### ISSUE TITLE
Feature request: sleep() function

### ISSUE THREAD

**Issue by IvanVeloz** [blocksds/sdk#156]
To complement the growing set of POSIX functions, a sleep() and usleep() functions could be added to BlocksDS, making use of a hardware timer as a monotonic clock.

I understand using up timers this can interfere with some games, but it's so handy for applications (and since we have the monotonic clock, we can also implement a bunch of other time functions). In my opinion, it should be available by default under ` unistd.h`, with an option to be disabled by something like `#define NO_MONOTONIC`.

**Comment by AntonioND** [comment]
It seems like a reasonable request. We already have some POSIX functions that work in a relatively special way. I guess the implementation could check if there is any free timer and use that one, or we could hardcode a timer, or even have a helper function that can switch the timer that is used.

I don't expect almost any game to use `sleep()` for anything, so I guess it doesn't have a big impact outside of helping port Linux applications to BlocksDS.

We would need to document the final behaviour of the function too.

I'm open to suggestions.

**Comment by asiekierka** [comment]
I think this should wait until at least one broader thing is considered: a system for allocating and freeing hardware resources like timers, DMA, etc., that homebrew in general can be encouraged to use, so we don't fall into conflicts.

**Comment by asiekierka** [comment]
Really, at that point we should consider morphing the coroutine system into some kind of RTOS - but I don't have the necessary field experience to start that conversation.

**Comment by IvanVeloz** [comment]
I see what you mean. I don't have that experience writing OSes either. I think in the meantime I'm going to write it as an external library with the POSIX API. That way the source code is available if we ever get there.

**Comment by AntonioND** [comment]
Yes, I think that you having your own implementation of `sleep()` is the best idea for now.

**Comment by IvanVeloz** [comment]
Sounds good! I'm closing the feature request as "not planned".

**Comment by asiekierka** [comment]
That is appreciated, but we still want to get around to this at some point; just not soon.

**Comment by AntonioND** [comment]
I'm thinking that we could just have a workaround like `timerAssignToLibc(int timer_id)`. If this function isn't called, `sleep()` and similar will crash with `libndsCrash()`. If the function has been called, `sleep()` uses the timer assigned to libc.

In the hypothetical case that a more complex system is implemented in libnds in the future we could just turn `timerAssignToLibc()` into a no-op.
